grpc-gateway/Dockerfile

```diff
@@ -0,0 +1,42 @@
+FROM ubuntu:19.10
+
+#RUN add-apt-repository ppa:longsleep/golang-backports
+RUN apt-get update && \
+    apt-get install -y wget unzip htop \
+    openjdk-13-jdk \
+    golang-go \
+    git
+
+# Install protoc
+ENV PROTOC_VERSION=3.11.4
+RUN wget https://github.com/google/protobuf/releases/download/v${PROTOC_VERSION}/protoc-${PROTOC_VERSION}-linux-x86_64.zip \
+    -O /protoc-${PROTOC_VERSION}-linux-x86_64.zip && \
+    unzip /protoc-${PROTOC_VERSION}-linux-x86_64.zip -d /usr/local/ && \
+    rm -f /protoc-${PROTOC_VERSION}-linux-x86_64.zip
+
+RUN mkdir -p /code/output/
+ENV BASEPATH /code
+ENV OUTPATH /code/output
+
+COPY / ${BASEPATH}/
+WORKDIR  $BASEPATH
+
+ENV JAVA_TOOL_OPTIONS=-Dfile.encoding=UTF8
+RUN ./gradlew clean && ./gradlew installDist
+
+RUN mkdir /go
+ENV GOPATH /go
+ENV PATH $GOPATH/bin:/usr/local/go/bin:$PATH
+
+RUN mkdir -p "$GOPATH/src" "$GOPATH/bin" && chmod -R 777 "$GOPATH"
+#install required go protoc plugins to build grpc-gateway server
+RUN go get -u \
+    github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway \
+    github.com/grpc-ecosystem/grpc-gateway/protoc-gen-swagger \
+    github.com/golang/protobuf/protoc-gen-go
+
+#run protoc plugin for protoc-gen-go
+RUN protoc -I/code/src/main/proto -I/code/build/extracted-protos/main -I/code/build/extracted-include-protos/main --plugin=protoc-gen-go=$GOPATH/bin/protoc-gen-go --go_out=plugins=grpc:$OUTPATH /code/src/main/proto/analysis.proto /code/src/main/proto/luceneserver.proto /code/src/main/proto/search.proto
```

Comment:
Do we need to keep the generated files around ? If we are generating them in Docker during the build, why do we need to keep them in the repo ?

Reply:
The purpose of this Dockerfile is to generate 
1. grpc-gateway reverse proxy source code and
2. swagger json files 
These are copied back into src/main/proto once the task buildGrpcGateway is run.

We regenerate them on each run because there could be changes in the src/main/proto/luceneserver.proto thereby needing that source code and swagger.json to be updated.

We keep them in the repo because it 
1. it allows easy access to swagger spec
2. can run REST server out of the box by running `go run http_wrapper <grpc_port> <rest_port>`
3. Seems like it is not a very uncommon thing to keep generated go code in the repo for example:
https://github.com/grpc-ecosystem/grpc-gateway/tree/master/examples/proto/examplepb
https://github.com/elastic/cloud-on-k8s/blob/master/pkg/apis/elasticsearch/v1/zz_generated.deepcopy.go

Reply:
1. I agree with keeping the generated swagger spec, but not sure about the go code
2. Don't get it, don't we still have to run `./gradlew clean installDist` to build the java protos and the jar ? We can generate go files in the same step I believe.
3. Yes, because of go's structure it makes it easy to import the generated code if you keep it. But assuming we're not going to actively work with the go code, not sure if we need to keep them.

If we do keep the files, does it make sense to keep them in a separate package rather than in src/main/proto ?

Reply:
> We can generate go files in the same step I believe.

`./gradlew installDist`  outputs the code needed to run java-grpc. This command runs fairly quickly.
`./gradlew buildGrpcGateway` generates the code needed to run rest over grpc. This has a lot deps including go deps like plugins and takes longer time.
When running travis build we run both commands as we need to provide both servers. Idea behind decoupling these two commands and not running them as a single command is:
1. easier developer experience. It is much faster to iterate when changing proto java objects and testing java grpc server only.
2. We might not always want to add every new grpc endpoint to be exposed via REST so we dont need to rebuild  the go server and reverse proxy in case a dev is only interested in modifying the grpc java files.  However eventually  we run builds on travis we will ensure that the correct go server is built.

Overall, I feel there is more benefit in keeping the building of go code separate from java code.
Also I do not see any harm in keeping the generated go code in the repo, since it is a pretty common practice, allows devs to take the rest server for a quick spin, with not many drawbacks of doing so.

> But assuming we're not going to actively work with the go code, not sure if we need to keep them
    Most go files except `http_wrapper.go` are auto generated. For the wrapper code I plan to make changes to it potentially like supporting running REST on same port as grpc.

Reply:
Alright. What about 
`If we do keep the files, does it make sense to keep them in a separate package rather than in src/main/proto ?`

This can also be done during the refactor though.

Reply:
moved to folder `grpc-gateway`